src/layers/python.rs

```diff
@@ -0,0 +1,582 @@
+        let cached_metadata = &layer_data.content_metadata.metadata;
+        let new_metadata = self.generate_layer_metadata(&context.stack_id);
+
+        if let Some(reason) = cache_invalidation_reason(cached_metadata, &new_metadata) {
+            log_info(format!("Discarding cache {reason}"));
+            Ok(ExistingLayerStrategy::Recreate)
+        } else {
+            log_info(format!(
+                "Using cached Python {}",
+                cached_metadata.python_version
+            ));
+            let PackagingToolVersions {
+                pip_version,
+                setuptools_version,
+                wheel_version,
+            } = &cached_metadata.packaging_tool_versions;
+            log_info(format!(
+                "Using cached pip {pip_version}, setuptools {setuptools_version} and wheel {wheel_version}"
+            ));
+            Ok(ExistingLayerStrategy::Keep)
+        }
+    }
+}
+
+impl<'a> PythonLayer<'a> {
+    fn generate_layer_metadata(&self, stack_id: &StackId) -> PythonLayerMetadata {
+        PythonLayerMetadata {
+            stack: stack_id.clone(),
+            python_version: self.python_version.to_string(),
+            packaging_tool_versions: self.packaging_tool_versions.clone(),
+        }
+    }
+}
+
+/// Metadata stored in the generated layer that allows future builds to determine whether
+/// the cached layer needs to be invalidated or not.
+#[derive(Clone, Deserialize, PartialEq, Serialize)]
+pub(crate) struct PythonLayerMetadata {
+    stack: StackId,
+    python_version: String,
+    packaging_tool_versions: PackagingToolVersions,
+}
+
+/// Compare cached layer metadata to the new layer metadata to determine if the cache should
+/// be invalidated, and if so, for what reason.
+fn cache_invalidation_reason(
+    cached_metadata: &PythonLayerMetadata,
+    new_metadata: &PythonLayerMetadata,
+) -> Option<String> {
+    // By destructuring here we ensure that if any additional fields are added to the layer
+    // metadata in the future, it forces them to be used as part of cache invalidation,
+    // otherwise Clippy would report unused variable errors.
+    let PythonLayerMetadata {
+        stack: cached_stack,
+        python_version: cached_python_version,
+        packaging_tool_versions:
+            PackagingToolVersions {
+                pip_version: cached_pip_version,
+                setuptools_version: cached_setuptools_version,
+                wheel_version: cached_wheel_version,
+            },
+    } = cached_metadata;
+
+    let PythonLayerMetadata {
+        stack,
+        python_version,
+        packaging_tool_versions:
+            PackagingToolVersions {
+                pip_version,
+                setuptools_version,
+                wheel_version,
+            },
+    } = new_metadata;
+
+    let mut reasons = Vec::new();
+
+    if cached_stack != stack {
+        reasons.push(format!(
```

Comment:
I really like this approach of building up all the reasons it's changed (rather than limiting to only one). I would love to come up with some way to make this easier/automagical for users via some trait or macro or something. Maybe manuel (or you) might have some ideas on interface and implementation.

It seems most of us have converged on similar patterns, it would be nice to make the common/default case require little to no code.

Reply:
> That said, I quite like having human readable explanations for the attributes, at which point any abstraction here would have to take that into account too (and so any abstraction may be unwieldy and not worth doing).

My philosophy is: We won't know until we try. Implementation is iteration. 

If we could metaprogram and dynamically get the field names we could use a convention of a default explanation based on the field name. Something like:

```
Detected changes, clearing the cache:
- Stack heroku-18 changed to heroku-22
- Python version 2.7 changed to 3.0
```

I think we could maybe codegen something via a macro which would (should?) have access to field names.

Of course we could still all write our own bespoke comparisons and messages, but it would be nice to have the option of a really solid/easy default, even if it's not perfect.